### DISCUSSION TITLE
Can I get a signal within PHP HTTP worker on destroying intention?

### DISCUSSION THREAD

**Discussion by 7krasov-casafari** [roadrunner-server/roadrunner#1362] [Q&A] [answered]
Hi guys,

I'm searching for a way how to gracefully finish the PHP worker on destroying/killing intention (e.g. close connections etc)

Imagine I have such a configuration

```
http:
  address: 0.0.0.0:8082
  pool:
    num_workers: 1
    max_jobs: 20
    destroy_timeout: 10s
```

When a worker reaches `max_jobs`=20, the RR kills it and creates a new one. I'm trying to catch some signal within PHP (pcntl_signal function) but it never runs. I see there is `destroy_timeout` option but it seems I understand it wrong because my worker is being closed immediately anyway despite this parameter setting
I would know also if some other reason can send me a signal (e.g. RR server got some signal from the container that is going to be reloaded/stopped, not just max_jobs is reached)

**rr version 2.10.7 (build time: 2022-07-14T09:02:31+0000, go1.18.4), OS: linux, arch: amd64**

Here is an example of an interception of signals

```
<?php

declare(strict_types=1);
declare(ticks=1);

namespace Pcntl;

use Log\Logger;
use Cycle\Database\DatabaseInterface;

class PcntlSignalsRegistrar
{
    public function register(DatabaseInterface $database, Logger $logger): void
    {
        pcntl_async_signals(true);
        $handlerClosure = static function (int $sigNum) use ($database, $logger) {
            $logger->warning(sprintf('pcntl_signal handler is called. SIG_NUM: %d', $sigNum));
            try {
                if ($database->getDriver()->isConnected()) {
                    $logger->warning('DB connection is opened. Going to close it gracefully');
                    if ($database->getDriver()->getTransactionLevel()) {
                        $logger->warning('DB connection is in transaction. Rolling back...');
                        $database->getDriver()->rollbackTransaction();
                    } else {
                        $logger->warning('DB connection is not in transaction');
                    }
                    $logger->warning('Performing a  DB disconnect...');
                    $database->getDriver()->disconnect();
                    return;
                }
                $logger->warning('DB connection is not opened, nothing to close');
            } catch (\Throwable $t) {
                $logger->logThrowable($t, 'Unable to gracefully shutdown a DB connection');
            }
        };
        foreach ([SIGHUP, SIGINT, SIGQUIT, SIGTERM, SIGUSR1, SIGUSR2] as $sigNum) {
            pcntl_signal($sigNum, $handlerClosure);
        }
    }
}

```

Thank you!

**Comment by rustatian** [comment, marked as answer]
Hey @7krasov-casafari 👋🏻 
On destroy, RR sends a [stop command](https://github.com/spiral/roadrunner-worker/blob/master/src/Worker.php#L175) via our protocol. We don't send signals. So, the worker finishes gracefully, allowing all destructors to run.

But I guess you should update to the latest version to have that.

**Reply by 7krasov** [reply]
Thanks for the information @rustatian :bow: 
Tried the newest version **rr version 2.11.4 (build time: 2022-10-06T14:42:36+0000, go1.19.2), OS: linux, arch: amd64**. 

I see that the condition in a worker _cycle while_ gets **false** value (this value is obtained from code, the place of the link you provided above :point_up: ) and the cycle ends without a worker killing 
```
while ($req = $workerPsr->waitRequest()) {
}
```

So I can palace a code for connection closing after this cycle and it has 10 seconds to perform this and this is great :+1: 

But, the only docker container restarting triggers it, but not max worker jobs reached event. The latest case seems to kill a worker immediately so my connection closing code is never reached

Is it expected behavior? Can we be sure that "max jobs reached" event triggers a stop signal sending by RR instead of immediate killing?

Thank you!

**Reply by rustatian** [reply]
I double-checked that case, and yes, you're right, the `max jobs event` was skipped in the 2.11.4 by mistake. Could you please update to the 2.12.0-rc.1 and try that again?

**Reply by 7krasov** [reply]
Yeah, I can confirm that both container reloading and max job event trigger stop signal sending  now :slightly_smiling_face: :muscle: 
Thank you! When are you planning a new release?

**Reply by rustatian** [reply]
Nice 😄 
I'll release RR on the 24th of November during our monthly SpiralScout Ecosystem News event 😃 Feel free to join ⚡

**Reply by 7krasov** [reply]
Sure, I will test a new release, thanks for the rapid support :rocket:  :smile: !

**Reply by 7krasov** [reply]
Hi @rustatian !
I have just tested on a new release. It works well for both "max jobs reached" event and the external signal, thank you! :bow:

**Reply by rustatian** [reply]
Great 👍🏻 Enjoy RR 🤗
